File: scripts/generate-argos-draft.js
'use strict';

const fs = require('node:fs/promises');
const path = require('node:path');
const { execFile } = require('node:child_process');
const { promisify } = require('node:util');
const run = promisify(execFile);

const pages = ['index.html', 'download.html', 'about.html', 'privacy-policy.html', 'terms-conditions.html', 'refund-policy.html', 'licenses.html'];
const locale = process.argv[2];
if (!locale) throw new Error('Pass a locale code, e.g. fr');

const attrs = /\b(?:aria-label|alt|title|placeholder|content)=(['"])([\s\S]*?)\1/gi;
const opaque = /<script\b[\s\S]*?<\/script\s*>|<style\b[\s\S]*?<\/style\s*>/gi;
const protectedTerms = ['HushBook', 'App Store', 'Google Play', 'HushBook Engine', 'LibriVox', 'Internet Archive', 'Sentry', 'RevenueCat'];
const decode = s => s.replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"').replace(/&#39;/g, "'");
const encode = s => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
function translatable(s) { const v = decode(s).replace(/\s+/g, ' ').trim(); return v && /[A-Za-z]/.test(v) && !/^https?:|^\/|^#|^mailto:|^[\w.-]+@[\w.-]+$/.test(v) && !/^[©®™—–·•]+$/.test(v); }
function mask(s) { const saved=[]; let out=s; for (const term of protectedTerms) { const token=`HBTERM${saved.length}X`; if(out.includes(term)){saved.push([token,term]);out=out.split(term).join(token);} } return [out,saved]; }
function unmask(s,saved){ for(const [token,term] of saved)s=s.split(token).join(term); return s; }
async function translate(s) { const [masked,saved]=mask(s); const {stdout}=await run('argos-translate.exe',['--from-lang','en','--to-lang',locale,masked],{windowsHide:true}); return unmask(stdout.trim(),saved); }

async function main(){
  const dir=path.join(process.cwd(),locale); await fs.mkdir(dir,{recursive:true});
  for(const page of pages){
    let html=await fs.readFile(page,'utf8'); const values=new Set();
    const visible=html.replace(opaque,'').replace(/<[^>]+>/g,'|');
    for(const part of visible.split('|')) if(translatable(part)) values.add(part.trim());
    for(const m of html.matchAll(attrs)) if(translatable(m[2])) values.add(m[2]);
    const map=new Map(); for(const value of values){ try{map.set(value,await translate(value));}catch(e){console.warn(`${locale} ${page}: ${value.slice(0,50)} (${e.message})`);} }
    html=html.replace(opaque, x=>`\u0000${Buffer.from(x).toString('base64')}\u0000`);
    html=html.replace(/>([^<>]+)</g,(all,raw)=>{const key=raw.trim(); return map.has(key)?`>${raw.replace(key,encode(map.get(key)))}<`:all;});
    html=html.replace(attrs,(all,q,raw)=>{const key=raw.trim(); return map.has(key)?all.replace(raw,encode(map.get(key))):all;});
    html=html.replace(/\u0000([A-Za-z0-9+/=]+)\u0000/g,(_,b)=>Buffer.from(b,'base64').toString('utf8'));
    html=html.replace(/<html lang="en">/i,`<html lang="${locale}">`);
    await fs.writeFile(path.join(dir,page),html,'utf8'); console.log(`${locale}: ${page} (${map.size} strings)`);
  }
}
main().catch(e=>{console.error(e);process.exitCode=1;});

File: scripts/generate_argos_draft.py
import html, re, sys
from pathlib import Path
import argostranslate.translate as tr

pages = ['index.html','download.html','about.html','privacy-policy.html','terms-conditions.html','refund-policy.html','licenses.html']
locale = sys.argv[1]
output_locale = sys.argv[2] if len(sys.argv) > 2 else locale
langs = tr.get_installed_languages(); source = next(x for x in langs if x.code == 'en'); target = next(x for x in langs if x.code == locale)
translator = source.get_translation(target)
attr = re.compile(r'\b(?:aria-label|alt|title|placeholder|content)=(\"|\')(.*?)\1', re.I|re.S)
opaque = re.compile(r'<script\b.*?</script\s*>|<style\b.*?</style\s*>', re.I|re.S)
protected = ['HushBook','App Store','Google Play','HushBook Engine','LibriVox','Internet Archive','Sentry','RevenueCat']
cache = {}
def can(s):
    s = html.unescape(s).strip()
    return bool(s and re.search('[A-Za-z]', s) and not re.match(r'^(https?:|/|#|mailto:|[\w.-]+@[\w.-]+$)', s))
def translate(s):
    key = html.unescape(s).strip()
    if key in cache: return cache[key]
    saved=[]; out=key
    for i,t in enumerate(protected):
        token=f'HBTERM{i}X'
        if t in out: saved.append((token,t)); out=out.replace(t,token)
    value=translator.translate(out)
    for token,t in saved: value=value.replace(token,t)
    cache[key]=value
    return value
def esc(s): return html.escape(s, quote=False)
outdir=Path(output_locale); outdir.mkdir(exist_ok=True)
for page in pages:
    text=Path(page).read_text(encoding='utf-8')
    opaque_blocks=[]
    def hide(m): opaque_blocks.append(m.group(0)); return f'HBOPAQUE{len(opaque_blocks)-1}X'
    work=opaque.sub(hide,text)
    values=[]
    for m in re.finditer(r'>([^<>]+)<', work):
        if can(m.group(1)) and html.unescape(m.group(1)).strip() not in values: values.append(html.unescape(m.group(1)).strip())
    for m in attr.finditer(work):
        if can(m.group(2)) and html.unescape(m.group(2)).strip() not in values: values.append(html.unescape(m.group(2)).strip())
    for start in range(0,len(values),24):
        batch=values[start:start+24]
        marker='\nHBSEP9X\n'
        translated=translator.translate(marker.join(batch)).split(marker)
        if len(translated)==len(batch):
            for source,value in zip(batch,translated): cache[source]=value.strip()
        else:
            for source in batch: translate(source)
    def node(m):
        raw=m.group(1); key=html.unescape(raw).strip()
        return m.group(0) if not can(raw) else '>'+raw.replace(key,esc(translate(raw)))+'<'
    work=re.sub(r'>([^<>]+)<',node,work)
    def attribute(m):
        raw=m.group(2); return m.group(0) if not can(raw) else m.group(0).replace(raw,html.escape(translate(raw),quote=True))
    work=attr.sub(attribute,work)
    for i,b in enumerate(opaque_blocks): work=work.replace(f'HBOPAQUE{i}X',b)
    work=re.sub(r'<html lang="en">',f'<html lang="{locale}">',work,flags=re.I)
    (outdir/page).write_text(work,encoding='utf-8'); print(output_locale,page,len(cache),flush=True)
